File: frontend/src/app/core/service/permission-service.ts
import { inject, Injectable } from '@angular/core';
import { PermissionApiService } from '@data/api/permission-api-service';
import { ApiErrorDetail } from '@data/interfaces/common-interface';
import { PageableParams } from '@data/interfaces/pageable-interface';
import { Permission, PermissionRequest } from '@data/interfaces/permission-interface';
import { catchError, throwError } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class PermissionService {
  private permissionApiService = inject(PermissionApiService);

  create(permissionData: PermissionRequest) {
    return this.permissionApiService.create(permissionData)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  findById(id: number) {
    return this.permissionApiService.findById(id)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  findAll(pageableParams: PageableParams = {}) {
    return this.permissionApiService.findAll(pageableParams)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  update(id: number, permissionData: PermissionRequest) {
    return this.permissionApiService.update(id, permissionData)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  delete(id: number) {
    return this.permissionApiService.delete(id)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }
}

File: backend/src/main/java/com/sdrouet/easy_restaurant/service/PermissionService.java
package com.sdrouet.easy_restaurant.service;

import com.sdrouet.easy_restaurant.dto.permission.PermissionRequestDTO;
import com.sdrouet.easy_restaurant.dto.permission.PermissionResponseDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface PermissionService {
    PermissionResponseDTO create(PermissionRequestDTO requestDTO);

    PermissionResponseDTO findById(Long id);

    Page<PermissionResponseDTO> findAll(Pageable pageable);

    PermissionResponseDTO update(Long id, PermissionRequestDTO requestDTO);

    void delete(Long id);
}
